### ISSUE TITLE
incorrect attribute / stats points from various ways of changing levels (e.g. FCV) - need generic function to do allocation / deallocation

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#6482]
Staff/Admins: Adjust anything here as you see fit. 

The bugs described here happen only under non-standard conditions and aren't critical to fix. I recommend that we don't work on this issue until after the API v3 changes are live.

A player's allocated plus unallocated attribute points should add up to their current level (or to 100 if their level is over 100, with exceptions listed below). Under some circumstances, this does not occur - e.g., from past bugs or from using Fix Character Values to adjust level. Once the points have got out of sync with the level, they aren't always brought back into sync during levelling up or dying.

To fix this, we could have a generic function that performs the following actions:
- adds up the allocated and unallocated attribute points
- compares them to the current level OR to 100 if the level is over 100
- if there are too few points, EITHER:
  - run the `autoAllocate` function if the user has chosen auto-allocation
  - add unallocated attribute points (`stats.points`)
- if there are too many points:
  - first subtract from the unallocated points, as much as necessary or until the unallocated points are zero
  - if still too many points AND if the user is not above level 100, subtract from allocated points, choosing the affected stat(s) in the same manner they are currently chosen after death
- never let any attribute have negative points
- never let unallocated points be negative

That function could then be run under the following circumstances:
- levelling up
- death
- rebirth
- reset
- Fix Character Values when the level is changed
- class change
- anything else?

There are special conditions for players who are above level 100, and special treatment for those who already have more than 100 points. Note the decision that is documented in "Attribute Points Change" in [Skills Rebalancing and Other Changes](http://habitica.wikia.com/wiki/User_blog:LadyAlys/Skills_Rebalancing_and_Other_Changes):
_"Players who already have attribute points above 100 will not have those points removed."_
- If they level up, they are not given a point unless their points are less than 100.
- If they rebirth or reset, their points drop to 1 as for any other player, and from that time on, they cannot ever gain more than 100 points.
- If they do a class change, their points drop to 100 and from that time on, they cannot ever gain more than 100 points.
- If they die and they have **un**allocated points, they lose all those unallocated points. If they die and have no unallocated points, they lose only **one** allocated point.
- Under any other circumstances, if they happen to have more **allocated** points than 100, the function described above does **not** reduced them to 100.

We allow players over level 100 to have more than 100 points solely because of the decision to let them keep their points when the stats/skills rebalancing occurred in early 2015. However any code that allowed them to keep excess points would also allow newer players to keep excess points if they got those points due to future bugs, and that is undesirable. I'm thinking that maybe we should put a flag in the accounts of any players now who are over 100 and have more than 100 points. The function could then use that flag to apply the special treatment, allowing all new players to be treated normally above level 100.

**Comment by DrStrangepork** [comment]
> I'm thinking that maybe we should put a flag in the accounts of any players now who are over 100 and have more than 100 points.

Wouldn't that require a new key-pair to be added to all user accounts? A universal boolean?

**Comment by Alys** [comment]
@DrStrangepork More or less. A missing flag would be treated as a false flag.

**Comment by DrStrangepork** [comment]
Ok, then wouldn't all the 100+ people need a key pair added to their user accounts? It would be either a universal boolean of True/False for the key value or an "if exist" check of the key. In either case, the 100+ set of people would need the key to exist in their user account. And if that is true, then wouldn't that key pair addition be a prerequisite to this issue?

**Comment by Alys** [comment]
@DrStrangepork Yes, that's correct, unless the person who works on this comes up with a different solution.

**Comment by DrStrangepork** [comment]
You could just add the logic directly, but then you'd be calculating it at every circumstance for everyone. Skipping on a boolean would be slightly faster for the (vast) majority of users who are not in the 100+ crowd.

I can take a stab at this. But I think we'd need to define what the additional key pair would be in order to test against it and exempt anyone not in the 100+ crowd.

**Comment by crookedneighbor** [comment]
Before any code changes, I think we should do an audit of affected active users. That will tell us if this is worth doing.

**Comment by Alys** [comment]
More problems with attribute points:

Il Mago Grigio (test account) ( ac056eb4-eaca-4b53-ab8d-dfd43d42ee6a ):
After various reports of strange things with attribute points at low levels, I have created a test account and fast forwarded between the first few levels. Well, here my observations:
- at level 1-9 I have already points to allocate (1 per level), but there are no notifications about them.
- I have received the modal for unlocking the drop system at level 3 and again at level 5, 6, 7.
- At level 10 I have opted out, and I find myself with 19 allocated points, the 10 to STR as it should be when opting out, and 9 additionals that I have allocated to INT when levelling up at levels 1-9.
- For a sync error, I go down one level and I now have 9 in STR and 9 in INT. But I am still out of the class system.
- Levelling up though I don't get new points to allocate.
- Level 14, I have 18 allocated points, I choose a class, I now have 14 allocable points and a notification about them.


Il Mago Grigio (test account 2) ( dd95201f-ac81-48c8-a28a-a8052e151b70 ):
I have tried again with an other test account, this time I didn't opt out and went like this:
- level 1-9, i have allocable attribute points.
- level 10, I choose a class, but differently then before the points already allocated are not unallocated.

**Comment by Alys** [comment]
We've just found a user whose details I won't release here who currently has these allocated and unallocated Stat Points:
```
    "stats": {
        "points": 216608,
        "str": -232972,
        "con": 29000,
        "int": -41536,
        "per": 29000,
    }
```
That's not supposed to be possible. I'm about to fix their data since they'll probably be getting unexpected behaviour or possibly causing dramatic changes to their party members' Stats if they're casting buffs. I can give the User ID to other admins if they want it. We should look into this when making the generic function that this issue talks about, but I don't see it as being high priority. I don't think it's common.

**Comment by Alys** [comment]
I'm increasing the priority of this to medium. I've been seeing several reports of Stat Points (previously called Attribute Points) being incorrect lately so a bug has crept in somewhere that lets you get more than your level allows. An example is below.

I'm also going to look into how many players are affected by this from the top post:
> There are special conditions for players who are above level 100, and special treatment for those who already have more than 100 points. Note the decision that is documented in "Attribute Points Change" in Skills Rebalancing and Other Changes

I'll report back about whether I think we can get away with removing that grandfathering once I have a sense of how many users are affected.

_From geolojess💎 (da27bf23-48f7-469d-9a32-58521f9c84a9):_
_"(Android) I seem to have 6 extra attribute points? I'm level 104, so I should have all 100 points already, but it seems I have 6 extra🤔"_

Even though that's on Android, if we fix this in the API, the fix will be good for all clients and third-party tools.

**Comment by Alys** [comment]
In my previous post I said I would report back about whether I think we can get away with removing that grandfathering. I believe we can. 

There's about 250 players who have more than 105 Points. I've ignored those between 100 and 105 Points because they could easily have got the extras just from the recent bugs; maybe players with more than 105 Points have too.

There's about 140 players with more than 120 Points.

If you count only players who logged in in the past few days, there's only 90 with more than 120 points.I think that's a low enough number of players for us to go back on our promise from 2015 and say that the new code will remove ALL Stat Points above 100 as soon as a player takes any action that affects their level (including leveling UP) or as soon as a player allocates one or more Points. Doing so will make the code simpler, with less place for bugs to hide.

**Comment by Alys** [comment]
There's a report of another bug related to Stats Points, from cTheDragons (bc10ec41-4213-4482-bbb8-558e3c3ea8d5):

"Have recently change class to Mage which unallocated all my attribute (stat) points. I have it to allocate automatically by class. And although I have increased by level 2 times. The unallocate points are still there."

**Comment by lemoness** [comment]
I agree that we can simplify the code to remove the grandfathering given the situation.

**Comment by D10master** [comment]
Can I try to work on it?
